### PR TITLE
[nn] InstanceNorm : no batch dim for modules

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: #60585 

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### CODE REVIEW

**Review by jbschlosser — CHANGES_REQUESTED** [review]
Thanks for the update! I think we actually want to support no batch dim at the op level, since we want the functional form to support it as well (for functorch purposes). I realize InstanceNorm calls BatchNorm underneath, but in theory I think this should be feasible haha

**Comment by kshitij12345** [comment]
I am not sure how to proceed about the same for functional.

The functional variant is not limited to a particular shape so it treats the first dim as the batch-dim. I don't think we can determine whether user meant batched or unbatched input from the shape for this function.

Thanks @jbschlosser!

https://github.com/pytorch/pytorch/blob/f83250fd4e1427f978a1d013620276d3c6e0ec0a/aten/src/ATen/native/Normalization.cpp#L584-L600

**Comment by jbschlosser** [comment]
> I am not sure how to proceed about the same for functional.
> 
> WIth the current API, the functional variant is not limited to a particular shape so it treats the first dim as the batch-dim. I don't think we can determine whether user meant batched or unbatched input from the shape for this function.
> 
> Thanks @jbschlosser!
> 
> https://github.com/pytorch/pytorch/blob/f83250fd4e1427f978a1d013620276d3c6e0ec0a/aten/src/ATen/native/Normalization.cpp#L584-L600

I see- this is a problem because `instance_norm` handles 1D/2D/3D cases in the same function. I guess this is why `conv1d`, `conv2d`, and `conv3d` exist as separate ops. Thanks for pointing this out - I'll think about how to handle this when I get back from PTO.

**Comment by jbschlosser** [comment]
> hi~ I have question about torch.nn.instancenorm; The mean/var shape is [C] not [N,C], why? I find that instancenorm kernel is used batchnorm’s kernel in cuda, the mean/var of Bn is C ,so the mean/var of instancenrom is [C] too? [https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Normalization.cpp#L616](url) `// we alias running_mean and running_var because they are const but we want to modify their data if (running_mean.defined()) { at::alias(running_mean).copy_(running_mean_.view({ b, c }).mean(0, false)); } if (running_var.defined()) { at::alias(running_var).copy_(running_var_.view({ b, c }).mean(0, false)); }`

@wei61547-jp `instance_norm()` on an input of shape `(N, C, H, W)` runs `batch_norm()` underneath on the input reshaped to `(1, N * C, H, W)`. This follows the diagram [here](https://i.stack.imgur.com/1JdN6.png) demonstrating computation of the various norms, effectively normalizing across the height and width dimensions per (batch item, channel). This gives a running mean / var output of shape `(N * C)`.

The running mean / var are then viewed / averaged across the batch to get shape C, as you mentioned above. If the running mean / var were left as shape `(N * C)` or `(N, C)`, this would require every batch sent through `instance_norm()` to be the same size. So the running mean / var are maintained as per-channel values to be generalizable across input batch sizes.

**Comment by kshitij12345** [comment]
Ping @jbschlosser. Module only updates are ready in this PR.

File: torch/csrc/api/include/torch/nn/modules/instancenorm.h [hunk 1]
```diff
@@ -10,14 +10,38 @@ namespace nn {
...
     this->_check_input_dim(input);
-    return torch::nn::functional::detail::instance_norm(
-      input, this->running_mean, this->running_var, this->weight, this->bias,
-      this->is_training() || !this->options.track_running_stats(), this->options.momentum(), this->options.eps());
+
+    // check if input does not have a batch-dim
+    // For InstanceNorm1D, 2D is unbatched and 3D is batched
+    // For InstanceNorm2D, 3D is unbatched and 4D is batched
+    // For InstanceNorm3D, 4D is unbatched and 5D is batched
+    if (input.dim() == D + 1) {
```
Comment (jbschlosser) [inline torch/csrc/api/include/torch/nn/modules/instancenorm.h:40]: I think this should actually be D + 2 due to the channels dim.

Also I think we need some C++-side tests to catch things like this, since the definition is separate across python / C++.

File: torch/nn/modules/instancenorm.py [hunk 2]
```diff
@@ -279,9 +287,13 @@ class LazyInstanceNorm2d(_LazyNormBase, _InstanceNorm):
 
     cls_to_become = InstanceNorm2d  # type: ignore[assignment]
 
+    def _get_no_batch_dim(self):
+        return 3
+
     def _check_input_dim(self, input):
-        if input.dim() != 4:
-            raise ValueError("expected 4D input (got {}D input)".format(input.dim()))
+        if input.dim() not in (3, 4):
```
Comment (jbschlosser) [inline torch/nn/modules/instancenorm.py:294]: Also need doc shape updates for `LazyInstanceNorm2d` above

File: torch/nn/modules/instancenorm.py [hunk 3]
```diff
@@ -114,8 +127,8 @@ class InstanceNorm1d(_InstanceNorm):
             statistics in both training and eval modes. Default: ``False``
 
     Shape:
-        - Input: :math:`(N, C, L)`
-        - Output: :math:`(N, C, L)` (same shape as input)
+        - Input: :math:`(N, C, L)` or :math:`(C, L)`
```
Comment (jbschlosser) [inline torch/nn/modules/instancenorm.py:130]: `num_features` doc needs to be updated above. It falsely claims that an input of shape `(N, L)` is possible. Currently, an error is thrown for a 2D input, and after this PR, a 2D input will be treated as `(C, L)`.

Same change is needed for the `LazyInstanceNorm1d` docs.

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! Couple minor comments below

File: torch/csrc/api/include/torch/nn/modules/instancenorm.h (hunk 1 above)
Comment (kshitij12345) [reply torch/csrc/api/include/torch/nn/modules/instancenorm.h:40]: This is actually correct, we are checking for the unbatched case here. I will add comment here.

Also, this is tested with C++ Parity test. (Changed it to D+2 and it failed locally).

Comment (jbschlosser) [reply torch/csrc/api/include/torch/nn/modules/instancenorm.h:40]: Awesome - thanks for clarifying! Looks good then :)

**Review by jbschlosser — APPROVED** [review]
LGTM! Thanks for the update :)
